drivers/tty/serdev/pibridge.c

```diff
@@ -0,0 +1,409 @@
+	mutex_unlock(&pi->lock);
+
+	wake_up_interruptible(&pi->read_queue);
+
+	if (ret < count)
+		pb_err(25, "put received buffer to fifo error(count:%d, ret:%d)\n", count, ret);
+	return ret;
+}
+
+static const struct serdev_device_ops pibridge_serdev_ops = {
+	.receive_buf	= pibridge_receive_buf,
+	.write_wakeup	= serdev_device_write_wakeup,
+};
+
+static int pibridge_parse_dt(struct serdev_device *serdev)
+{
+	struct device_node *node = serdev->dev.of_node;
+	u32 speed = 115200;
+
+	of_property_read_u32(node, "current-speed", &speed);
+	serdev_device_set_baudrate(serdev, speed);
+
+	/* RTS is used to drive Transmit Enable pin, hence no flow control */
+	serdev_device_set_flow_control(serdev, false);
+
+	return serdev_device_set_parity(serdev, SERDEV_PARITY_EVEN);
+}
+
+
+static int pibridge_probe(struct serdev_device *serdev)
+{
+	struct device *dev = &serdev->dev;
+	struct pibridge *pi;
+	int ret;
+
+	pi = devm_kzalloc(dev, sizeof(*pi), GFP_KERNEL);
+	if (!pi)
+		return -ENOMEM;
+
+	pibridge_s = pi;
+	pi->serdev = serdev;
+
+	serdev_device_set_drvdata(serdev, pi);
+	serdev_device_set_client_ops(serdev, &pibridge_serdev_ops);
+
+	mutex_init(&pi->lock);
+	init_waitqueue_head(&pi->read_queue);
+
+	ret = kfifo_alloc(&pi->read_fifo, REV_PI_RECV_BUFFER_SIZE, GFP_KERNEL);
+	if (ret)
+		return ret;
+
+	ret = serdev_device_open(serdev);
+	if (ret)
+		goto err_kfifo_free;
+
+	ret = pibridge_parse_dt(serdev);
+	if (ret)
+		goto err_serdev_close;
+
+	return 0;
+
+err_serdev_close:
+	serdev_device_close(serdev);
+err_kfifo_free:
+	kfifo_free(&pi->read_fifo);
+	return ret;
+}
+
+static void pibridge_remove(struct serdev_device *serdev)
+{
+	struct pibridge *pi = serdev_device_get_drvdata(serdev);
+
+	serdev_device_close(serdev);
+
+	kfifo_free(&pi->read_fifo);
+};
+
+#ifdef CONFIG_OF
+static const struct of_device_id pibridge_of_match[] = {
+	{ .compatible = "kunbus,pibridge" },
+	{},
+};
+MODULE_DEVICE_TABLE(of, pibridge_of_match);
+#endif
+
+static struct serdev_device_driver pibridge_driver = {
+	.driver	= {
+		.name		= "pibridge",
+		.of_match_table	= of_match_ptr(pibridge_of_match),
+	},
+	.probe	= pibridge_probe,
+	.remove	= pibridge_remove,
+};
+module_serdev_device_driver(pibridge_driver);
+
+
+/*****************/
+
+int pibridge_send(u8 *buf, u16 len)
+{
+	struct pibridge *pi = pibridge_s;
+	struct serdev_device *serdev = pi->serdev;
+	int ret;
+
+	ret = serdev_device_write(serdev, buf, len, MAX_SCHEDULE_TIMEOUT);
+	if (ret < 0)
+		pb_err(25, "write to serdev error(len:%d)\n", len);
+
+	/* read fifo may contain stale bytes */
+	mutex_lock(&pi->lock);
+	kfifo_reset(&pi->read_fifo);
+	mutex_unlock(&pi->lock);
+
+	serdev_device_wait_until_sent(serdev, 0);
+
+	return ret;
+}
+
+int pibridge_recv_timeout(u8 *buf, u16 len, u16 timeout)
+{
+	struct pibridge *pi = pibridge_s;
+	int jiffies;
+	int ret;
+
+	jiffies = wait_event_timeout(pi->read_queue, kfifo_len(&pi->read_fifo) >= len,
+			msecs_to_jiffies(timeout));
+	mutex_lock(&pi->lock);
+	if (jiffies){
```

Comment:
I would not call the current code wrong, since we specified a timeout and therefore can treat the expiration as an error. But yes, your suggestion would be a slight improvement of the current implementation (also in the current implementation we do not have to take the mutex before the "if" condidtion).